staging/src/k8s.io/apimachinery/pkg/util/proxy/dial.go

```diff
@@ -87,46 +83,29 @@ func dialURL(ctx context.Context, url *url.URL, transport http.RoundTripper) (ne
 				tlsConfig = tlsConfigCopy
 			}
 
-			// Since this method is primary used within a "Connection: Upgrade" call we assume the caller is
+			// Since this method is primarily used within a "Connection: Upgrade" call we assume the caller is
 			// going to write HTTP/1.1 request to the wire. http2 should not be allowed in the TLSConfig.NextProtos,
 			// so we explicitly set that here. We only do this check if the TLSConfig support http/1.1.
 			if supportsHTTP11(tlsConfig.NextProtos) {
 				tlsConfig = tlsConfig.Clone()
 				tlsConfig.NextProtos = []string{"http/1.1"}
 			}
 
-			tlsConn = tls.Client(netConn, tlsConfig)
-			if err := tlsConn.Handshake(); err != nil {
+			tlsConn := tls.Client(netConn, tlsConfig)
+			if err := tlsConn.HandshakeContext(ctx); err != nil {
 				netConn.Close()
 				return nil, err
 			}
-
+			return tlsConn, nil
 		} else {
-			// Dial. This Dial method does not allow to pass a context unfortunately
-			tlsConn, err = tls.Dial("tcp", dialAddr, tlsConfig)
-			if err != nil {
-				return nil, err
+			// Dial.
+			tlsDialer := tls.Dialer{
```

Comment:
https://pkg.go.dev/crypto/tls#Dialer admits a custom dialer, I don't know if you can replace all the `https` block but

```
tlsDialer := tls.Dialer{
  Config: tlsConfig,
  NetDialer: dialer,
}
```